### PR TITLE
Add lifecycle hooks option for keycloak container

### PR DESCRIPTION
**Pull request by radumazi**
Signed-off-by: radu mazilu <radumazi2@gmail.com>

#### What this PR does / why we need it:
Adds container lifecycle hooks for the keycloak chart

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Special notes for your reviewer:
This is my first pull request on github. I went through the documentation but if I missed anything please point it out. Thank you
I'm not sure where I should @-mention the maintainers so I'm doing it here: @unguiculus @thomasdarimont

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/keycloak/templates/statefulset.yaml [hunk 1]
```diff
@@ -71,6 +71,10 @@ spec:
           imagePullPolicy: {{ .Values.keycloak.image.pullPolicy }}
           command:
             - /scripts/keycloak.sh
+          {{- if .Values.keycloak.lifecycleHooks }}
+          lifecycle:
+{{ tpl .Values.keycloak.lifecycleHooks . |indent 12 }}
+          {{- end }}
```
Comment (unguiculus) [inline stable/keycloak/templates/statefulset.yaml]: Please change the formatting to match the existing style:

```yaml
        {{- if .Values.keycloak.lifecycleHooks }}
          lifecycle:
{{ tpl .Values.keycloak.lifecycleHooks . | indent 12 }}
        {{- end }}
```

File: stable/keycloak/values.yaml [hunk 2]
```diff
@@ -43,6 +43,12 @@ keycloak:
   ## Custom script that is run before Keycloak is started.
   preStartScript:
 
+  ## lifecycleHooks defines the container lifecycle hooks
+  lifecycleHooks: |
+    #postStart:
```
Comment (unguiculus) [inline stable/keycloak/values.yaml]: Add space after `#`

File: stable/keycloak/Chart.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 name: keycloak
-version: 4.8.0
+version: 4.8.1
```
Comment (unguiculus) [inline stable/keycloak/Chart.yaml]: Bump minor version for new feature.

File: stable/keycloak/README.md [hunk 4]
```diff
@@ -69,6 +69,7 @@ Parameter | Description | Default
 `keycloak.hostAliases` | Mapping between IP and hostnames that will be injected as entries in the pod's hosts files | `[]`
 `keycloak.securityContext` | Security context for the pod | `{runAsUser: 1000, fsGroup: 1000, runAsNonRoot: true}`
 `keycloak.preStartScript` | Custom script to run before Keycloak starts up | ``
+`keycloak.lifecycleHooks` | Container lifecycle hooks. Passed through the `tpl` function and thus to be configured a string | `""
```
Comment (unguiculus) [inline stable/keycloak/README.md]:
```suggestion
`keycloak.lifecycleHooks` | Container lifecycle hooks. Passed through the `tpl` function and thus to be configured a string | ``
```

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks @radumazi. Would you mind sharing what your use case is for this?

**Comment by radumazi** [comment]
@unguiculus I made the required modifications. My use case is that I want to export metrics from keycloak in order for prometheus to be able to query them. For that I used this service provider https://github.com/aerogear/keycloak-metrics-spi that can be enabled either from the interface or from the cli, and as I'm building an automated pipeline for deploying keycloak I needed to do it from the cli.

To do so I use these commands found in the service provider docs:
`/opt/jboss/keycloak/bin/kcadm.sh config credentials --server http://localhost:8080/auth --realm master --user $KEYCLOAK_USER --password $KEYCLOAK_PASSWORD
/opt/jboss/keycloak/bin/kcadm.sh update events/config -s "eventsEnabled=true" -s "adminEventsEnabled=true" -s "eventsListeners+=metrics-listener"
/usr/bin/rm -f /opt/jboss/.keycloak/kcadm.config`

Keycloak must be up and running before I use these commands, so my solution was to write a script that runs at the postStart container lifecycle hook and waits for the connection to :8080 to become available, then runs the commands.

If you know of a better way to do this, feel free to correct me

Thank you

**Comment by radumazi** [comment]
@unguiculus As I told you, I'm unfamiliar with the process, so I hope it's not a stupid question. Is there anything I more I can do for this pull request? I understand that the lgtm and approved labels needs to be added. Thank you

**Comment by unguiculus** [comment]
@radumazi Please bare with us. I wanted to have @thomasdarimont to have a look as well.

**Comment by radumazi** [comment]
@unguiculus I just wasn't sure if there was anything else I should have done, cause it still displayed "changes required". Anyhow, thank you, and have a nice day!

**Pull request merged**